libs/pika/runtime/tests/unit/CMakeLists.txt

```diff
@@ -23,22 +23,153 @@ foreach(test ${tests})
+)
+
+# Disable binding through environment variable, but enable it through command line option. The
+# latter should take precedence.
+pika_add_configuration_precedence_test(
+  configuration_precedence 4 2 "--pika:print-bind --pika:bind=balanced"
+  "${balanced_binding_expected_output}" "PIKA_BIND=none"
+)
+
+string(
+  CONCAT
+    second_pu_binding_one_thread_expected_output
+    "   0: PU L#[0-9]+\\(P#1\\), Core L#[0-9]+\\(P#[0-9]+\\)(, Socket L#[0-9]+\\(P#[0-9]+\\))?(, NUMANode L#[0-9]+\\(P#[0-9]+\\))?(, Socket L#[0-9]+\\(P#[0-9]+\\))?, on pool \"default\"\n"
+    "All tests passed."
+)
+string(CONCAT no_binding_one_thread_expected_output
+              "   0: thread binding disabled, on pool \"default\"\n" "All tests passed."
+)
+
+# Set process mask through command line option.
+if(APPLE)
+  set(configuration_precedence_expected_output "${no_binding_one_thread_expected_output}")
+else()
+  set(configuration_precedence_expected_output "${second_pu_binding_one_thread_expected_output}")
+endif()
+pika_add_configuration_precedence_test(
+  configuration_precedence 5 1 "--pika:print-bind --pika:process-mask=0x2"
+  "${configuration_precedence_expected_output}" ""
+)
+
+# Set process mask through environment variable.
+if(APPLE)
+  set(configuration_precedence_expected_output "${no_binding_one_thread_expected_output}")
+else()
+  set(configuration_precedence_expected_output "${second_pu_binding_one_thread_expected_output}")
+endif()
+pika_add_configuration_precedence_test(
+  configuration_precedence 6 1 "--pika:print-bind" "${configuration_precedence_expected_output}"
+  "PIKA_PROCESS_MASK=0x2"
+)
+
+# Set process mask through environment variable and command line option. The latter should take
+# precedence.
+if(APPLE)
+  set(configuration_precedence_expected_output "${no_binding_one_thread_expected_output}")
+else()
+  set(configuration_precedence_expected_output "${second_pu_binding_one_thread_expected_output}")
+endif()
+pika_add_configuration_precedence_test(
+  configuration_precedence 7 1 "--pika:print-bind --pika:process-mask=0x2"
+  "${configuration_precedence_expected_output}" "PIKA_PROCESS_MASK=0x4"
+)
+
+string(CONCAT one_thread_expected_output "   0:.*\n" "All tests passed.")
+string(CONCAT two_threads_expected_output "   0:.*\n" "   1:.*\n" "All tests passed.")
+
+# Set number of threads through command line option.
+pika_add_configuration_precedence_test(
+  configuration_precedence 8 -3 "--pika:print-bind --pika:threads=1"
+  "${one_thread_expected_output}" ""
+)
+
+# Set number of threads through environment variable.
+pika_add_configuration_precedence_test(
+  configuration_precedence 9 -3 "--pika:print-bind" "${two_threads_expected_output}"
+  "PIKA_THREADS=2"
+)
+
+# Set number of threads through environment variable and command line option.
+pika_add_configuration_precedence_test(
+  configuration_precedence 10 -3 "--pika:print-bind --pika:threads=2"
+  "${two_threads_expected_output}" "PIKA_THREADS=1"
+)
```

Comment:
This is not particularly pretty. Nicer ways of doing this or is this ok?

Are there (important) combinations of options missing that are not tested?